plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/ui/LoginBrowser.java

```diff
@@ -0,0 +1,96 @@
+package com.google.cloud.tools.eclipse.appengine.login.ui;
+
+import java.net.URL;
+
+import org.eclipse.jface.dialogs.Dialog;
+import org.eclipse.jface.dialogs.IDialogConstants;
+import org.eclipse.jface.layout.GridDataFactory;
+import org.eclipse.jface.window.IShellProvider;
+import org.eclipse.swt.SWT;
+import org.eclipse.swt.browser.Browser;
+import org.eclipse.swt.browser.ProgressAdapter;
+import org.eclipse.swt.browser.ProgressEvent;
+import org.eclipse.swt.browser.TitleEvent;
+import org.eclipse.swt.browser.TitleListener;
+import org.eclipse.swt.widgets.Composite;
+import org.eclipse.swt.widgets.Control;
+import org.eclipse.swt.widgets.Shell;
+
+public class LoginBrowser extends Dialog {
+
+  private static final String LOGOUT_URL = "https://www.google.com/accounts/Logout"; //$NON-NLS-1$
+  private static final String SUCCESS_CODE_PREFIX = "Success code="; //$NON-NLS-1$
+
+  private Browser browser;
+  private URL loginUrl;
+
+  private String verificationCode;
+
+  public LoginBrowser(IShellProvider shellProvider, URL loginUrl) {
+    super(shellProvider.getShell());
```

Comment:
Since you're immediately dereferencing the `IShellProvider` to get a `Shell` you may as well just take a `Shell`.  `IShellProvider` is more useful when you want to defer the obtaining of the `Shell` until a later time.